### DISCUSSION TITLE
Rewrite the core

### DISCUSSION THREAD

**Discussion by Steelskin** [visualboyadvance-m/visualboyadvance-m#1295] [Ideas]
# Introduction

In order to get the Link feature working reliably, with netplay, I believe we first need to modernize the codebase. VBA-m was originally written and designed in C89, and it is showing its age. It is not a bad codebase, but it is old. I believe the only way to achieve good results with Link netplay is to emulate every instance locally, the way Dolphin now handles netplay with GBA emulation. To achieve this, we need to break the amount of global state and ensure every bit of data has clear ownership.

At the same time, more changes mean potentially more breakage. We need more testing and a testing strategy. However, the amount of global state makes this much harder. To build unit test binaries, we would need to incorporate pretty much the entirety of the core emulator. This complicates the task of breaking up the global state.

This discussion is a proposal for a course of action to get rid of the global state and start rewriting parts of the core in a more modern manner. Some sections present several options and their advantages and drawbacks, with the aim of fostering discussions to get to a decision.

# Areas of discussion

## 1. Moving to explicit ownership

We want to get into a state where a single emulator instance is comprised of a set of data that is entirely owned by a single object. This will allow us to start multiple emulator instances within the same project, on separate threads. More broadly, this would also improve the wx frontend experience by allowing us to run even a single emulator instance on a separate thread, avoiding many of the pauses that arise from opening the menus, for instance.

The long term plan is to have objects that handle specific aspects of emulation. Off the top of my mind, I can think of the following objects: `GB[A]Memory`, `GB[A]CPU`, `GB[A]PPU`, `GB[A]Options`, `CoreOptions`, `CoreAudio`, `CoreLink`
(the audio and link emulation are entangled together), owned by their respective `GB[A]Emulator`.

Moving to explicit ownership can be done piecemeal:
1. **Introduce structures to hold what is currently global variables**
   Instead of having a lot of global variables, we use global structs. For every global variable that we find, we put it in a global struct.
2. **Move functions to the new structs**
   Build around the variables we moved and move global functions to structs. Keeping standalone utility functions is fine, but emulator-related exported functionality should be move to the structs.
3. **Narrow down the scope of the variables**
   Start moving variables to be privately owned by the structs, rather than being part of the public API.

This should also be done for the wx frontend. Effort is currently underway to move every non-const global object to be owned by `wxvbamApp`. However, most classes should access the owned data via a `*Provider` API, similar to this:

```c++
namespace config {

// Forward declarations.
class Bindings

class ConfigProvider {
   // Global getters/setters.
   static ConfigProvider* Get();
   static void Set(ConfigProvider* config_provider);

   virtual ~ConfigProvider() = default;

   virtual Bindings* bindings() = 0;
};

} // namespace config
```

This will ease testability down the road. Since these objects will not directly depend on the `wxvbamApp` declaration, test harnesses will be able to provide a `ConfigProvider` object, without needing to depend on the entirety of the wx frontend and the core.

## 2. Rewriting the core

As we will need to rewrite large parts of the code to do the above work, it is worth considering how we want the core to work down the line.

### Objectives

* Cleaner, simpler API
* Builds completely separately from the frontends
* Explicit ownership
* Thoroughly tested

### Option 2.A: Rust

One option is to rewrite the core in Rust, bit by bit. Rust is a recent, popular memory-safe language with mature tooling.

#### Advantages

**Sum types**

Sum types are tagged unions with their own syntax in Rust, derived from the ML family of languages. They allow to easily express and use complex data types with a few lines of code. For instance, consider the following type that could be used to represent a GBA cheat code:

```rust
enum GBACheat {
    GameShark(String),
    Generic(String),
    CodeBreaker(String),
    ActionReplay(String),
}
```

With a just a single type, we can easily define both the type of cheat and its value.

**Better error reporting**

Rust provides a generic (Rust equivalent of C++ templates) `Result` type that can be either a result value or an error. It is roughly defined this way:

```rust
enum Result<T, E> {
   Ok(T),
   Err(E),
}
```

This type is used all over the Rust ecosystem and makes it much easier to deal with errors in a uniform manner. Coupled with sum types, we can get a much more explicit error, by having specific types for return values. Following from the previous example, we can express the error for decoding a cheat code with this sum type:

```rust
enum CheatsDecodeError {
   /// The provided string does not have the expected length.
   InvalidLength,
   /// The character at the specified index is invalid.
   InvalidCharacter(usize),
   /// The address part of the cheat is not a hexadecimal value.
   AddressNotHex,
   /// The value part of the cheat is not a hexadecimal value.
   ValueNotHex,
}
```

Finally, this also allows us to easily remove localized error messages from the core, keeping it smaller and moving that responsibility to the embedder.

**Integrated testing framework**

Rust comes with its own build system (Cargo), which includes its own testing support. This makes it much easier to integrate tests within a Rust project than within a C++ project. This also encourages better development practices.

#### Risks

**C++ Interop**

We need to interop with C++, this means generating C++ bindings for our Rust core. From my research, there is currently no good way of handling this. Rust internal FFI (Foreign Function Interface) mechanism is first and foremost designed for C interop.

Some crates (the Rust ecosystem equivalent of libraries) provide some C++ bindings but I have found all of them to be lacking in one way or the other. The [`cxx`](https://cxx.rs/) crate, for instance, does not properly converts the `Result` type to a C++ union type, but converts it to 

I believe the best way to work around the issue is to build our own interop system based on an Interface Definition Language, which will generate bindings for both Rust and C++. To reduce the complexity of this work, we can simply
focus on calling Rust from C++, not the other way around. In general, it is relatively painless to export C functions around a C++ class, and Rust can easily call these, as needed. Though, I believe this will be seldom-needed as I would suggest taking a bottoms-up approach in rewriting the core, starting from the lowest levels, up to the core API.

**Console targets**

Since we still need to support console targets, we will also need the Rust code to be built for these platforms. The risk here is minimal, but we are currently lacking build coverage for these targets on CI. The Rust toolchains exist for these targets, but whether or not they are functional is an open question.

Furthermore, the libretro build currently uses plain old Makefiles. Integrating these with Cargo is another challenge.

**Integration with CMake**

This is a very minor risk but it is still worth mentioning. We will need to integrate the Cargo build with CMake. There are CMake modules to do so and I do not expect this to be a huge problem. I have a prototype to make this work. The Rust and C++ parts can use different toolchains since the Rust project will output a static library.

### Option 2.B: Interface Definition Language (IDL)

We can define an abstract API in an IDL and generate bindings for the core on one side and the frontends on the other. This allows us to keep a clear and clean abstraction between the core and the frontends.

As mentioned in the previous section, this may be required if we go with the Rust option.

#### Advantages

**Generate bindings for multiple languages**

We would not be limited to C++ for the frontend. Given that the IDL itself is language-agnostic, it would be a "simple" matter of generating bindings for the appropriate language.

**Possibility of moving the core emulator out of process**

For security reasons, it would be best to move the core emulator execution out of process, in a sandboxed process, following modern web browser design. While the risk remains low, ROM hacks are a popular use case for the emulator and many of them are not "real" GBA software in that they rely on emulator quirks to work, and fail to work properly on real hardware. These are more or less dangerous to run as they rely on a number of unintentional and undocumented side effects of the emulator.

Using an IDL as the core API would allow us to achieve this more easily, since it would then be a matter of having a library and generating bindings on either end. For prior art, see [protobuf](https://protobuf.dev/), [Mojo](https://chromium.googlesource.com/chromium/src/+/master/mojo/README.md) or [FIDL](https://fuchsia.dev/fuchsia-src/get-started/learn/fidl/fidl).

#### Risks

**Initial development cost**

We will need to design and develop an IDL and a binding generator before we can decide on the core API. More likely than not, these would be develop hand-in-hand with an initial IDL covering basic use cases and being expanded upon following the needs of the core emulator.

### Option 2.C: C++ classes

With this option, we move the core from being mostly C with a bit of C++ to be almost entirely C++, with modern C++ design practices.

#### Advantages

**No preliminary work**

By and large, this is the most straightforward option as it does not require large preliminary work before we can start making significant changes to the core design.

## 3. Testing strategy

Working in a well-tested codebase makes it much easier to land new changes since breakages are caught very quickly, even before they land.

Tests should be run on CI for every PR submitted before landing to the master branch. This is currently not the case.

## 3.i. C++ Test framework

We need to decide on a test framework to use for our needs. For Rust code, this is not a problem, since Cargo comes with its own test facilities. However, there are multiple options for C++.

Considering the existing amount of testing, porting our existing tests to another framework is going to be a painless process.

### Option 3.i.A: [GoogleTest](https://github.com/google/googletest)

*Disclosure: I am a former Google employee. I have used GoogleTest extensively and also have contributions in GoogleTest.*

GoogleTest is Google's open source C++ test framework. It is used extensively within the company for both close source and open source projects.

#### Advantages

**Well-supported and maintained**

GoogleTest is as close as there is to an industry standard when it comes to C++ test frameworks. It is widely used throughout the industry, not just at Google. This ensures the project is going to keep being maintained for years to come with little to no action on our part.

**Feature-rich**

GoogleTest has a lot of features, especially compared to any of the alternatives. In particular, GoogleTest includes a mock support, which can come in handy to test that various events are properly fired or to easily mock part of an API that we don't fully control.

Another interesting feature is death tests, which can be used to test that a piece of software successfully crashes on assertion failure.

**Easy CMake integration**

GoogleTest can be [easily](https://google.github.io/googletest/quickstart-cmake.html) integrated with CMake. The main issue would be to convert our existing tests to GoogleTest but that too can be done quickly.

### Option 3.i.B: Keep using DocTest

We could keep using DocTest. This is the simpler option since it doesn't involve doing any changes to our build system. And while DocTest is a single header, its reduced feature set is going to be a problem down the line.

## 3.ii. Core unit tests

The core currently has zero internal tests. Individual parts of the core should come with their own tests. As we start rewriting the core, we should strive to add tests for any part for which we are doing large refactors.

## 3.iii. Test ROMs

In addition to the above, we should have test compliance ROMs and a way to check that we behave in the same manner as the original hardware.

I would argue that we should develop our own ROM test infrastructure where we can test aspects of the CPU/PPU/other emulation independently, with a way to automatically check that the test has succeeded.

## 3.iv. Frontend tests

The wx frontend also has very little testing and they are not run on CI. Similarly to the work being done in the core, we should make sure individual classes can be built independently from the rest of the system.

Sub-modules for the wx frontend should be their own individual targets and should be independently testable. Individual dialogs and widgets should be tested to ensure that various scenario are working properly and do not regress over time.

## 3.v. Reproducible builds

Our dependency management always pulls in the latest version of the dependencies. We should have tagged versions of our dependencies and be able to pull the exact same version that was used to build a specific revision in the first place.

This would be a great help in bisecting issues whenever a user reports a regression.

## 3.vi. Automatic crash reporting

There is currently no crash reporting mechanism. We should have the ability to dump a file on crash, so we can figure out where and why the emulator crashed. Longer term, it would be ideal for that file to be automatically uploaded to a server so we can track where the issues are.

## 3.vii Build coverage

Our build coverage on CI has improved a lot over the past months. We are still missing a couple of builds on CI:
* **Console libretro builds**
  We need to figure out a way to pull the toolchains and decide on which platforms we want to support. My recommendation is thus:
  * GameCube / Wii / Wii U
  * Switch
  * Playstation 3 / 4 / 5
  * Playstation Vita
  * Xbox 360 / One / Series
* **Official release builds**
  Official release builds use a different build process that is not documented and not part of CI coverage. This needs to be added.

## 3.viii. Better user bugs

Many users fail to properly fill the entirety of the bug template. I would argue that it should be easier to do so. Here are a few suggestions:

1. **Change the template order**
   We should put the emulator version, operating system, etc fields at the top of the template, *before* the user starts filling out the actual issue. This might entice more people to fully fill the bug template.
2. **Pre-fill the template from the emulator**
   All of the information in the template is already known to the emulator itself. My guess is most users get to the template from the "Report a bug" link under the "Help" menu. We should be able to provide the information directly in the link menu, maybe in the form of URL queries and use that to pre-fill the template.

## 4. Licensing for libraries

Depending on the options we decide to go with, some of the work described above may involve the creation of new libraries. Since we are starting new projects, we have the occasion to consider what kind of license we want. The only requirement is that the library licensing remains compatible with GPLv2 or later from the existing code base.

### Option 4.A: GPLv2 or later

This is the most straightforward option. However, it is going to reduce reusability of libraries outside of our needs since it is only possible to link statically against a GPL project and it has to be open source.

### Option 4.B: LGLPv2 or later

The LGPL offers the option for close source projects to link dynamically against an LGPL library. The main issue is the dynamic linking requirement, which comes with costs in terms of performance and binary size.

### Option 4.C: MPL 2.0

The MPL is compatible with GPLv2 or later code as well as explicitly allowing static linking against the MPL 2.0 code. The only requirement is that changes to the MPL 2.0 code be contributed back to the project.

### Option 4.D: A more permissive license (BSD, MIT, Apache)

Finally, one last option is to use a more permissive open source license that comes with little to no requirements from the embedder.

# Conclusion

My personal recommendation is to go with the following options:
* 2.A Rust
* 3.i.A GoogleTest
* 4.C MPL 2.0

I think these would allow us to make a truly great and modern emulator that can be reused by other projects but I understand that these are really large changes and would appreciate the input from the rest of the team.

**Comment by ZachBacon** [comment]
To be honest I'm onboard with the idea of rust. And with another project I'm with going to be adopting some tools that are custom written in rust (after originally being written in dlang) This would help me I feel personally adopt rust myself. Outside of that, it would definitely be a task. And with the interopt with c++ using cxx this will allow us to make that adoption easier, still fix what we need short term in c++ but slowly migrate to a rust codebase.
